python/ray/util/client/server/dataservicer.py

```diff
@@ -46,15 +81,30 @@ def __init__(self, basic_service: "RayletServicer"):
         self.basic_service = basic_service
         self.clients_lock = Lock()
         self.num_clients = 0  # guarded by self.clients_lock
+        # dictionary mapping client_id's to the last time they connected
+        self.client_last_seen: Dict[str, float] = {}
+        # dictionary mapping client_id's to their reconnect grace periods
+        self.reconnect_grace_periods: Dict[str, float] = {}
+        # dictionary mapping client_id's to their response cache
+        self.response_caches: Dict[str, OrderedResponseCache] = defaultdict(
+            OrderedResponseCache)
+        # stopped event, useful for signals that the server is shut down
+        self.stopped = threading.Event()
 
     def Datapath(self, request_iterator, context):
+        start_time = time.time()
+        # set to True if client shuts down gracefully
+        cleanup_requested = False
         metadata = {k: v for k, v in context.invocation_metadata()}
         client_id = metadata.get("client_id")
         if client_id is None:
             logger.error("Client connecting with no client_id")
             return
         logger.debug(f"New data connection from client {client_id}: ")
-        accepted_connection = self._init(client_id, context)
+        accepted_connection = self._init(client_id, context, start_time)
+        response_cache = self.response_caches[client_id]
+        # Set to False if client requests a reconnect grace period of 0
+        reconnect_enabled = True
```

Comment:
Looks like there are places this should be verified as well? Like in reconnect on the client?

Reply:
Like when you check if it can reconnect?

Reply:
On the client side this is handled all through [this line](https://github.com/ray-project/ray/blob/bbeba2d12e1d1aa8caedcd40793e320d12297c59/python/ray/util/client/worker.py#L242) (every function that attempts retries goes through _can_reconnect, i.e. logsclient, dataclient, and regular raylet servicer RPCs)